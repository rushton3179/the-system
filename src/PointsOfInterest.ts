import React from 'react';

/**
 * All the points of interest in the solar system that get rendered.
 * Values are obtained from either https://nssdc.gsfc.nasa.gov/planetary/factsheet/ or wikipedia.
 *
 * The following properties may or may not exist on each point of interest,
 * - testId: string - A unique string used for data-testid identifiers for tests.
 * - ref: RefObject<HTMLDivElement> - A ref so that we can scroll to the object when it is selected in the nav menu.
 * - display: string - The name of the point.
 * - radius: number - The radius of the point of interest (includes rings if visible), units are km.
 * - distance: number - The points distance from the sun, units are km.
 * - orbitalPeriod: number - How many days it takes the point to orbit the sun.
 * - Satellites: PointOfInterest[] - An array of satellites that orbit the point. These have radius, distance,
 *  orbitalPeriod just like the points.
 */
export const pointsOfInterest = {
  sun: {
    testId: 'sun',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Sun',
    radius: 695508,
    distance: 0,
    orbitalPeriod: 1,
  },
  mercury: {
    testId: 'mercury',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Mercury',
    radius: 2440,
    distance: 57900000,
    orbitalPeriod: 88,
  },
  venus: {
    testId: 'venus',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Venus',
    radius: 6052,
    distance: 108200000,
    orbitalPeriod: 224.7,
  },
  earth: {
    testId: 'earth',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Earth',
    radius: 6378,
    distance: 149600000,
    orbitalPeriod: 365.2,
    satellites: [
      {
        testId: 'moon',
        distance: 384400,
        orbitalPeriod: 27.3,
        radius: 1737,
      },
    ],
  },
  mars: {
    testId: 'mars',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Mars',
    radius: 3396,
    distance: 227900000,
    orbitalPeriod: 687,
  },
  theBelt: {
    ref: React.createRef<HTMLDivElement>(),
    display: 'The Belt',
  },
  jupiter: {
    testId: 'jupiter',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Jupiter',
    radius: 71498,
    distance: 778600000,
    orbitalPeriod: 4331,
    satellites: [
      {
        testId: 'io',
        distance: 421700,
        orbitalPeriod: 1.77,
        radius: 1822,
      },
      {
        testId: 'europa',
        distance: 670900,
        orbitalPeriod: 3.55,
        radius: 1561,
      },
      {
        testId: 'ganymede',
        distance: 1070400,
        orbitalPeriod: 7.15,
        radius: 2634,
      },
      {
        testId: 'callisto',
        distance: 1882700,
        orbitalPeriod: 16.69,
        radius: 2410,
      },
    ],
  },
  saturn: {
    testId: 'saturn',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Saturn',
    radius: 80000 + 58232, // making the ring radius clear
    distance: 1433500000,
    orbitalPeriod: 10747,
    satellites: [
      {
        testId: 'titan',
        distance: 1221870,
        orbitalPeriod: 15.95,
        radius: 2574,
      },
      {
        testId: 'rhea',
        distance: 527108,
        orbitalPeriod: 1.77,
        radius: 764,
      },
    ],
  },
  uranus: {
    testId: 'uranus',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Uranus',
    radius: 25559,
    distance: 2872500000,
    orbitalPeriod: 30589,
  },
  neptune: {
    testId: 'neptune',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Neptune',
    radius: 24764,
    distance: 4495100000,
    orbitalPeriod: 59800,
  },
  pluto: {
    testId: 'pluto',
    ref: React.createRef<HTMLDivElement>(),
    display: 'Pluto',
    radius: 1185,
    distance: 5906400000,
    orbitalPeriod: 90560,
  },
} as const;
