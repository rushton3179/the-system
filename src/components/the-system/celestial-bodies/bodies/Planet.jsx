import React from "react";
import PropTypes from "prop-types";
import SystemContext from "../../../../SystemContext";
import CelestialBody from "./CelestialBody";
import { SunConsts } from "../../../../SharedConsts";

const moonToCB = (moon, multipliers, planetRadius) => {
    const { className, radius, orbitalPeriod, distance } = moon;
    const { satelliteDist, orbitalPeriodMultiplier, sizeMultiplier } = multipliers;
    const moonRadius = radius * sizeMultiplier;

    return {
        className,
        radius: moonRadius,
        distance: distance * satelliteDist - planetRadius * sizeMultiplier - moonRadius,
        orbitalPeriod: orbitalPeriod * orbitalPeriodMultiplier,
        planetRadius
    };
};

const applyMultipliers = (consts, mults) => ({
    distance:
        consts.distance * mults.distanceMultiplier +
        (SunConsts.radius - consts.radius) * mults.sizeMultiplier,
    radius: consts.radius * mults.sizeMultiplier,
    orbitalPeriod: consts.orbitalPeriod * mults.orbitalPeriodMultiplier
});

const planet = (name, moons, planetConstants, scrollToRef, context) => {
    const { multipliers } = context;
    const { distance, radius, orbitalPeriod } = applyMultipliers(planetConstants, multipliers);

    const satellites = moons.map(moon => moonToCB(moon, multipliers, radius));

    return (
        <CelestialBody
            className={name}
            distance={distance}
            orbitalPeriod={orbitalPeriod}
            radius={radius}
            satellites={satellites}
            scrollToRef={scrollToRef}
        />
    );
};

const Planet = props => {
    const { name, moons, planetConstants, scrollToRef } = props;
    return (
        <SystemContext.Consumer>
            {context => planet(name, moons, planetConstants, scrollToRef, context)}
        </SystemContext.Consumer>
    );
};

Planet.propTypes = {
    name: PropTypes.string.isRequired,
    moons: PropTypes.arrayOf(PropTypes.shape({})),
    planetConstants: PropTypes.shape({
        radius: PropTypes.number.isRequired,
        distance: PropTypes.number.isRequired,
        orbitalPeriod: PropTypes.number.isRequired
    }).isRequired,
    scrollToRef: PropTypes.shape({})
};

Planet.defaultProps = {
    moons: [],
    scrollToRef: null
};

export default Planet;
