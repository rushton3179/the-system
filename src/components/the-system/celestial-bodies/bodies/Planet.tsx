import React, { useContext, RefObject, FC } from 'react';

import AppContext, { SystemMultipliers } from '../../../../SystemContext';
import CelestialBody, { CelestialBodyProps } from './CelestialBody';
import { SunConsts } from '../../../../SharedConsts';
import { MoonDetails, PlanetDetails } from '../../../../types';

const moonToCB = (moon: MoonDetails, multipliers: SystemMultipliers, planetRadius: number): CelestialBodyProps => {
  const { className, radius, orbitalPeriod, distance } = moon;
  const { satelliteDist, orbitalPeriodMultiplier, sizeMultiplier } = multipliers;
  const moonRadius = Math.max(0.5, radius * sizeMultiplier);

  return {
    className,
    radius: moonRadius,
    distance: distance * satelliteDist - planetRadius * sizeMultiplier - moonRadius,
    orbitalPeriod: orbitalPeriod * orbitalPeriodMultiplier,
    planetRadius,
    hasOrbitLine: true,
  };
};

const applyMultipliers = (consts: PlanetDetails, multipliers: SystemMultipliers): PlanetDetails => ({
  distance:
    consts.distance * multipliers.distanceMultiplier + (SunConsts.radius - consts.radius) * multipliers.sizeMultiplier,
  radius: consts.radius * multipliers.sizeMultiplier,
  orbitalPeriod: consts.orbitalPeriod * multipliers.orbitalPeriodMultiplier,
});

interface Props {
  name: string;
  planetConstants: PlanetDetails;
  scrollToRef: RefObject<HTMLDivElement>;
  moons?: MoonDetails[];
}

const Planet: FC<Props> = ({ name, moons, planetConstants, scrollToRef }) => {
  const { multipliers } = useContext(AppContext);
  const { distance, radius, orbitalPeriod } = applyMultipliers(planetConstants, multipliers);
  const satellites = moons?.map((moon: MoonDetails) => moonToCB(moon, multipliers, radius));

  return (
    <CelestialBody
      className={name}
      distance={distance}
      hasOrbitLine
      orbitalPeriod={orbitalPeriod}
      radius={radius}
      satellites={satellites}
      scrollToRef={scrollToRef}
    />
  );
};

export default Planet;
