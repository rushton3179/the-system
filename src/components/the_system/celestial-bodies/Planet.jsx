import React from "react";
import PropTypes from "prop-types";
import CelestialBody from "./CelestialBody";

const moonToCB = (moon, multipliers, planetRadius) => {
    const { className, radius, orbitalPeriod, distance } = moon;
    const { satelliteDist, orbitalPeriodMultiplier, sizeMultiplier } = multipliers;
    const moonRadius = radius * sizeMultiplier;

    return {
        className,
        radius: moonRadius,
        distance: distance * satelliteDist - planetRadius * sizeMultiplier - moonRadius,
        radiansPerMinute: orbitalPeriod * orbitalPeriodMultiplier,
        systemRadius: planetRadius
    };
};

const applyMultipliers = (consts, mults) => ({
    distance: consts.distance * mults.distanceMultiplier - consts.radius * mults.sizeMultiplier,
    radius: consts.radius * mults.sizeMultiplier,
    orbitalPeriod: consts.orbitalPeriod * mults.orbitalPeriodMultiplier
});

const Planet = props => {
    const { name, moons, planetConstants, multipliers, systemRadius } = props;
    const { distance, radius, orbitalPeriod } = applyMultipliers(planetConstants, multipliers);

    const satellites = moons.map(moon => moonToCB(moon, multipliers, radius));

    return (
        <CelestialBody
            className={name}
            distance={distance}
            radiansPerMinute={orbitalPeriod}
            radius={radius}
            satellites={satellites}
            systemRadius={systemRadius}
        />
    );
};

Planet.propTypes = {
    name: PropTypes.string.isRequired,
    moons: PropTypes.arrayOf(PropTypes.shape({})),
    multipliers: PropTypes.objectOf(PropTypes.number).isRequired,
    planetConstants: PropTypes.shape({
        radius: PropTypes.number.isRequired,
        distance: PropTypes.number.isRequired,
        orbitalPeriod: PropTypes.number.isRequired
    }).isRequired,
    systemRadius: PropTypes.number.isRequired
};

Planet.defaultProps = {
    moons: []
};

export default Planet;
