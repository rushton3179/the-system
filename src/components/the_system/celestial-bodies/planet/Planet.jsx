import React from "react";
import PropTypes from "prop-types";
import SystemContext from "../../SystemContext";
import CelestialBody from "../celestial-body/CelestialBody";

const moonToCB = (moon, multipliers, planetRadius) => {
    const { className, radius, orbitalPeriod, distance } = moon;
    const { satelliteDist, orbitalPeriodMultiplier, sizeMultiplier } = multipliers;
    const moonRadius = radius * sizeMultiplier;

    return {
        className,
        radius: moonRadius,
        distance: distance * satelliteDist - planetRadius * sizeMultiplier - moonRadius,
        radiansPerMinute: orbitalPeriod * orbitalPeriodMultiplier,
        planetRadius
    };
};

const applyMultipliers = (consts, mults) => ({
    distance: consts.distance * mults.distanceMultiplier - consts.radius * mults.sizeMultiplier,
    radius: consts.radius * mults.sizeMultiplier,
    orbitalPeriod: consts.orbitalPeriod * mults.orbitalPeriodMultiplier
});

const planet = (name, moons, planetConstants, context) => {
    const { multipliers } = context;
    const { distance, radius, orbitalPeriod } = applyMultipliers(planetConstants, multipliers);

    const satellites = moons.map(moon => moonToCB(moon, multipliers, radius));

    return (
        <CelestialBody
            className={name}
            distance={distance}
            radiansPerMinute={orbitalPeriod}
            radius={radius}
            satellites={satellites}
        />
    );
};

const Planet = props => {
    const { name, moons, planetConstants } = props;
    return (
        <SystemContext.Consumer>
            {context => planet(name, moons, planetConstants, context)}
        </SystemContext.Consumer>
    );
};

Planet.propTypes = {
    name: PropTypes.string.isRequired,
    moons: PropTypes.arrayOf(PropTypes.shape({})),
    planetConstants: PropTypes.shape({
        radius: PropTypes.number.isRequired,
        distance: PropTypes.number.isRequired,
        orbitalPeriod: PropTypes.number.isRequired
    }).isRequired
};

Planet.defaultProps = {
    moons: []
};

export default Planet;
