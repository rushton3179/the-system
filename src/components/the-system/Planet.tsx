import React, { useContext, RefObject, FC } from 'react';

import AppContext from '../../context/SystemContext';
import CelestialBody, { CelestialBodyProps } from './celestial-body/CelestialBody';
import { pointsOfInterest } from '../../PointsOfInterest';
import styles from './Planet.module.scss';

/** Represents details used for the moon of a planet. */
interface Satellite {
  /** The radius in km */
  radius: number;
  /** The distance from what it is orbiting in km */
  distance: number;
  /** Its orbital period in days. */
  orbitalPeriod: number;
}

interface Props {
  /** Class name to give the element, passed to CelestialBody. */
  className: string;
  /** A ref so that we can scroll to planet when it is selected in the nav menu. */
  scrollToRef: RefObject<HTMLDivElement>;
  /** The radius of the planet in km's. */
  radius: number;
  /** The distance from the planet to the Sun. */
  distance: number;
  /** The number of days the planet takes to orbit the sun */
  orbitalPeriod: number;
  /** Things that orbit the planet, all moons at this stage. */
  satellites?: readonly Satellite[];
}

/**
 * A planet in the solar system. This applies multipliers to values before they get passed to CelestialBody.
 */
const Planet: FC<Props> = ({ className, distance, radius, orbitalPeriod, scrollToRef, satellites }) => {
  const {
    systemRadius,
    multipliers: { distanceMultiplier, sizeMultiplier, orbitalPeriodMultiplier, satelliteDist },
  } = useContext(AppContext);

  const adjustedDistance = distance * distanceMultiplier + (pointsOfInterest.sun.radius - radius) * sizeMultiplier;
  const adjustedRadius = radius * sizeMultiplier;
  const adjustedOrbitalPeriod = orbitalPeriod * orbitalPeriodMultiplier;

  const satelliteToCelestialBody = (satellite: Satellite): CelestialBodyProps => {
    // Make sure the moons are at least 1px so they can be seen.
    const satelliteRadius = Math.max(1, satellite.radius * sizeMultiplier);

    return {
      className: styles.satellite,
      radius: satelliteRadius,
      distance: satellite.distance * satelliteDist + adjustedRadius - satelliteRadius,
      orbitalPeriod: satellite.orbitalPeriod * orbitalPeriodMultiplier,
      referenceRadius: adjustedRadius,
      hasOrbitLine: true,
    };
  };

  return (
    <CelestialBody
      className={className}
      distance={adjustedDistance}
      hasOrbitLine={true}
      orbitalPeriod={adjustedOrbitalPeriod}
      radius={adjustedRadius}
      satellites={satellites?.map(satelliteToCelestialBody)}
      referenceRadius={systemRadius}
      scrollToRef={scrollToRef}
    />
  );
};

export default Planet;
