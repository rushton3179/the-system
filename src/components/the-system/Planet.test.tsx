import { render, screen } from '@testing-library/react';
import React from 'react';

import { pointsOfInterest } from '../../PointsOfInterest';
import Planet from './Planet';

describe('Planet', () => {
  test('it renders the correct number of celestial bodies', () => {
    const { jupiter } = pointsOfInterest;
    render(
      <Planet
        className={'some-styles'}
        scrollToRef={jupiter.ref}
        radius={jupiter.radius}
        distance={jupiter.distance}
        orbitalPeriod={jupiter.orbitalPeriod}
        satellites={jupiter.satellites}
      />
    );
    expect(screen.getAllByTestId('celestial-body').length).toBe(1 + jupiter.satellites.length);
  });
});
