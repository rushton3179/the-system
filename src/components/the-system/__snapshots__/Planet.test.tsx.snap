// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Planet it matches the snapshot 1`] = `
<div
  class="orbit orbitLine"
  data-testid="undefined-orbit-line"
  style="animation: orbit 4331s linear infinite; height: 16267.508000000002px; left: 75888.8301142454px; top: 75888.8301142454px; width: 16267.508000000002px;"
>
  <div
    class="celestialBody some-styles"
    style="animation: planet-rotation 4331s linear infinite; height: 71.498px; left: -35.749px; top: 50%; width: 71.498px;"
    title="some-styles"
  >
    <div
      class="orbit orbitLine"
      data-testid="io-orbit-line"
      style="animation: orbit 1.77s linear infinite; height: 155.83800000000002px; left: -42.17000000000001px; top: -42.17000000000001px; width: 155.83800000000002px;"
    >
      <div
        class="celestialBody satellite"
        data-testid="io"
        style="animation: planet-rotation 1.77s linear infinite; height: 2px; left: -1px; top: 50%; width: 2px;"
        title="satellite"
      />
    </div>
    <div
      class="orbit orbitLine"
      data-testid="europa-orbit-line"
      style="animation: orbit 3.55s linear infinite; height: 205.678px; left: -67.09px; top: -67.09px; width: 205.678px;"
    >
      <div
        class="celestialBody satellite"
        data-testid="europa"
        style="animation: planet-rotation 3.55s linear infinite; height: 2px; left: -1px; top: 50%; width: 2px;"
        title="satellite"
      />
    </div>
    <div
      class="orbit orbitLine"
      data-testid="ganymede-orbit-line"
      style="animation: orbit 7.15s linear infinite; height: 285.57800000000003px; left: -107.04px; top: -107.04px; width: 285.57800000000003px;"
    >
      <div
        class="celestialBody satellite"
        data-testid="ganymede"
        style="animation: planet-rotation 7.15s linear infinite; height: 2.634px; left: -1.317px; top: 50%; width: 2.634px;"
        title="satellite"
      />
    </div>
    <div
      class="orbit orbitLine"
      data-testid="callisto-orbit-line"
      style="animation: orbit 16.69s linear infinite; height: 448.038px; left: -188.27px; top: -188.27px; width: 448.038px;"
    >
      <div
        class="celestialBody satellite"
        data-testid="callisto"
        style="animation: planet-rotation 16.69s linear infinite; height: 2.41px; left: -1.205px; top: 50%; width: 2.41px;"
        title="satellite"
      />
    </div>
  </div>
</div>
`;
